### PR TITLE
KAFKA-10892: Shared Readonly State Stores ( revisited )

### PR DESCRIPTION
**Pull request by calmera**
This PR implements KIP-813: support for shareable state stores.

Tests were created to validate the structure of the resulting topology.

Certain parts (MockProcessor eg.) have been migrated to the newer PAPI as well. Tests have been ran against the whole project to be sure this migration doesn't cause any side-effects.

The contribution is my original work and I license the work to the project under the project's open source license

This PR is a split of the previous PR #12188 which contained changes to the Topology as well as the PAPI test classes.

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/Topology.java [hunk 1]
```diff
@@ -737,6 +737,91 @@ public synchronized Topology addStateStore(final StoreBuilder<?> storeBuilder,
         return this;
     }
 
+    /**
+     * Adds a Read Only {@link StateStore} to the topology.
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/Topology.java]:
```suggestion
     * Adds a read-only {@link StateStore} to the topology.
```

File: streams/src/main/java/org/apache/kafka/streams/Topology.java [hunk 2]
```diff
@@ -737,6 +737,91 @@ public synchronized Topology addStateStore(final StoreBuilder<?> storeBuilder,
...
+     *
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/Topology.java]:
```suggestion
     * <p>
```

File: streams/src/main/java/org/apache/kafka/streams/Topology.java [hunk 3]
```diff
@@ -737,6 +737,91 @@ public synchronized Topology addStateStore(final StoreBuilder<?> storeBuilder,
...
+     * A Read Only StateStore can use any compacted topic as a changelog.
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/Topology.java]:
```suggestion
     * A read-only StateStore can use any compacted topic as a changelog.
```

File: streams/src/main/java/org/apache/kafka/streams/Topology.java [hunk 4]
```diff
@@ -737,6 +737,91 @@ public synchronized Topology addStateStore(final StoreBuilder<?> storeBuilder,
...
+     * <p>
+     * A {@link SourceNode} will be added to consume the data arriving from the partitions of the input topic.
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/Topology.java]: Seems to be too detailed. Also `SourceNode` is an internal class and I think we should not refer to it in the JavaDocs.

File: streams/src/main/java/org/apache/kafka/streams/Topology.java [hunk 5]
```diff
@@ -737,6 +737,91 @@ public synchronized Topology addStateStore(final StoreBuilder<?> storeBuilder,
...
+     * <p>
+     * The provided {@link ProcessorSupplier} will be used to create an {@link ProcessorNode} that will receive all
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/Topology.java]: Similar for `ProcessorNode`

File: streams/src/main/java/org/apache/kafka/streams/Topology.java [hunk 6]
```diff
@@ -737,6 +737,91 @@ public synchronized Topology addStateStore(final StoreBuilder<?> storeBuilder,
...
+     * records forwarded from the {@link SourceNode}.
+     * This {@link ProcessorNode} should be used to keep the {@link StateStore} up-to-date.
+     *
+     * @param storeBuilder          user defined key value store builder
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/Topology.java]: Why "key value" ?

File: streams/src/main/java/org/apache/kafka/streams/Topology.java [hunk 7]
```diff
@@ -737,6 +737,91 @@ public synchronized Topology addStateStore(final StoreBuilder<?> storeBuilder,
...
+     * @param sourceName            name of the {@link SourceNode} that will be automatically added
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/Topology.java:753]: as above

File: streams/src/main/java/org/apache/kafka/streams/Topology.java [hunk 8]
```diff
@@ -737,6 +737,91 @@ public synchronized Topology addStateStore(final StoreBuilder<?> storeBuilder,
...
+     * @param timestampExtractor    the stateless timestamp extractor used for this source,
+     *                              if not specified the default extractor defined in the configs will be used
+     * @param keyDeserializer       the {@link Deserializer} to deserialize keys with
+     * @param valueDeserializer     the {@link Deserializer} to deserialize values with
+     * @param topic                 the topic to source the data from
+     * @param processorName         the name of the {@link ProcessorSupplier}
+     * @param stateUpdateSupplier   the instance of {@link ProcessorSupplier}
+     * @return itself
+     * @throws TopologyException if the processor of state is already registered
+     */
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/Topology.java:763]: Should we mention anything about `auto.offset.reset` in the JavaDocs? Seems we don't allow to pass in a config, and we might want to hard-code it to "earliest"? (Cannot remember if/what we discussion on the KIP about it?)

File: streams/src/main/java/org/apache/kafka/streams/Topology.java [hunk 9]
```diff
@@ -737,6 +737,91 @@ public synchronized Topology addStateStore(final StoreBuilder<?> storeBuilder,
...
+    public synchronized <KIn, VIn> Topology addReadOnlyStateStore(final StoreBuilder<?> storeBuilder,
+                                                                  final String sourceName,
+                                                                  final TimestampExtractor timestampExtractor,
+                                                                  final Deserializer<KIn> keyDeserializer,
+                                                                  final Deserializer<VIn> valueDeserializer,
+                                                                  final String topic,
+                                                                  final String processorName,
+                                                                  final ProcessorSupplier<KIn, VIn, Void, Void> stateUpdateSupplier) {
+        if (storeBuilder.loggingEnabled()) {
+            // -- disabling logging. We might want to print some logging.
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/Topology.java]: For globlaKTable we also just disable logging without logging any warning -- guess we could mention it in the JavaDocs.

Also, we don't really need the `if` and can call `withLoggingDisabled()` blindly.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamBranchTest.java [hunk 10]
```diff
@@ -59,7 +59,7 @@ public void testKStreamBranch() {
 
         assertEquals(3, branches.length);
 
-        final MockProcessorSupplier<Integer, String> supplier = new MockProcessorSupplier<>();
+        final MockProcessorSupplier<Integer, String, Void, Void> supplier = new MockProcessorSupplier<>();
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamBranchTest.java]: Thanks for the cleanup!

File: streams/src/test/java/org/apache/kafka/test/MockProcessor.java [hunk 11]
```diff
@@ -28,9 +29,11 @@
 import java.util.List;
 import java.util.Map;
 
-@SuppressWarnings("deprecation") // Old PAPI. Needs to be migrated.
-public class MockProcessor<K, V> extends org.apache.kafka.streams.processor.AbstractProcessor<K, V> {
-    private final MockApiProcessor<K, V, Object, Object> delegate;
+public class MockProcessor<KIn, VIn, KOut, VOut> implements Processor<KIn, VIn, KOut, VOut> {
+    private final MockApiProcessor<KIn, VIn, KOut, VOut> delegate;
+
+    private ProcessorContext<KOut, VOut> context;
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/test/MockProcessor.java]: Why do we add this? We inherit `context` from the super class anyway.

File: streams/src/test/java/org/apache/kafka/streams/TopologyTest.java [hunk 12]
```diff
@@ -2303,6 +2304,63 @@ private void addGlobalStoreToTopologyAndExpectedDescription(final String globalS
         expectedDescription.addGlobalStore(expectedGlobalStore);
     }
 
+    @Test
+    public void readonlyStateStoresShouldHaveTheirOwnSubTopology() {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/TopologyTest.java]:
```suggestion
    public void readOnlyStateStoresShouldHaveTheirOwnSubTopology() {
```

File: streams/src/test/java/org/apache/kafka/streams/TopologyTest.java [hunk 13]
```diff
@@ -2303,6 +2304,63 @@ private void addGlobalStoreToTopologyAndExpectedDescription(final String globalS
...
+        allNodes.add(expectedSource);
+        allNodes.add(expectedProcessor);
+        expectedDescription.addSubtopology(new SubtopologyDescription(0, allNodes));
+
+        assertThat(topology.describe(), equalTo(expectedDescription));
+        assertThat(topology.describe().hashCode(), equalTo(expectedDescription.hashCode()));
+    }
+
+    @Test
+    public void readonlyStateStoresShouldNotLog() {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/TopologyTest.java]:
```suggestion
    public void readOnlyStateStoresShouldNotLog() {
```

File: docs/streams/developer-guide/processor-api.html [hunk 14]
```diff
@@ -396,6 +397,21 @@ <h2>
                 </ul>
                 </p>
             </div>
+            <div class="section" id="readonly-state-stores">
+                <span id="streams-developer-guide-state-store-readonly"></span><h3><a class="toc-backref" href="#id12">ReadOnly State Stores</a><a class="headerlink" href="#readonly-state-stores" title="Permalink to this headline"></a></h3>
+                <p>The changelog topic for a fault-tolerant state store can also be used to perform lookups against. It is
```
Comment (mjsax) [inline docs/streams/developer-guide/processor-api.html]: This sentence is somewhat confusing. Maybe we should start with something like this:

> A read-only state store materialized the data from its input topic. It also uses the input topic for fault-tolerance, and thus does not have an additional changelog topic (the input topic is re-used as changelog). Thus, the input topic should be configured with log compaction (add link to log compaction). Note that no other processor should modify the content of the state store, and the only writer should be the associated "state update processor"; other processor may read the content of the read-only store.

File: docs/streams/developer-guide/processor-api.html [hunk 15]
```diff
@@ -396,6 +397,21 @@ <h2>
...
+                    crucial to understand that there can be only one writer to this state store, but there can be multiple
+                    readers.</p>
+
+                <p>This allows you to use a readonly state store for looking up reference data managed by another processor
```
Comment (mjsax) [inline docs/streams/developer-guide/processor-api.html]:
```suggestion
                <p>This allows you to use a read-only state store for looking up reference data managed by another processor
```

File: docs/streams/developer-guide/processor-api.html [hunk 16]
```diff
@@ -396,6 +397,21 @@ <h2>
...
+                    or even another process all together. The messages within the compacted topic may be projected (subset of data extracted)
```
Comment (mjsax) [inline docs/streams/developer-guide/processor-api.html]: > may be projected 

I don't think that works? When we restore the state, we would by-pass the processor and put the bytes from the input topic into the store as-is...

File: docs/streams/developer-guide/processor-api.html [hunk 17]
```diff
@@ -396,6 +397,21 @@ <h2>
...
+                    before going into the readonly state store. Therefor we can optimize what the data within our readonly state
+                    store should look like, depending on the use-case it needs to serve.</p>
+
+                <p><b>note:</b> beware of the partitioning requirements when using readonly statestores for lookups during
```
Comment (mjsax) [inline docs/streams/developer-guide/processor-api.html]:
```suggestion
                <p><b>note:</b> beware of the partitioning requirements when using read-only state stores for lookups during
```

File: docs/streams/developer-guide/processor-api.html [hunk 18]
```diff
@@ -396,6 +397,21 @@ <h2>
...
+                    processing. You might want to make sure the original changelog topic is co-partitioned with the processors
+                    reading the readonly statestore.</p>
```
Comment (mjsax) [inline docs/streams/developer-guide/processor-api.html]:
```suggestion
                    reading the read-only state store.</p>
```

Comment (mjsax) [reply docs/streams/developer-guide/processor-api.html]: This reminds me: should we allow to pass in a custom partitioner? 🤔 (Could also be follow up work though...)

**Review by mjsax — COMMENTED** [review]
Thanks for the PR @calmera -- I finally got time to review.

Overall LGTM.

I think we should add a test using `TopologyTestDriver` to see if the data is really put into a state store and that we can access it via IQ and also via a second processor.

File: streams/src/main/java/org/apache/kafka/streams/Topology.java (hunk 9 above)
Comment (calmera) [reply streams/src/main/java/org/apache/kafka/streams/Topology.java]: agreed, will change accordingly

File: docs/streams/developer-guide/processor-api.html (hunk 16 above)
Comment (calmera) [reply docs/streams/developer-guide/processor-api.html]: yeah, pretty sure eventually we came to that conclusion as well. Too bad though, would have been nice for re-usability purposes.

I will change the description and that the projected stuff out.

File: streams/src/main/java/org/apache/kafka/streams/Topology.java (hunk 6 above)
Comment (calmera) [reply streams/src/main/java/org/apache/kafka/streams/Topology.java]: This indeed should just be any store builder

File: streams/src/main/java/org/apache/kafka/streams/Topology.java (hunk 7 above)
Comment (calmera) [reply streams/src/main/java/org/apache/kafka/streams/Topology.java:753]: I think this was copied from the description of `addGlobalStore`.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamBranchTest.java (hunk 10 above)
Comment (calmera) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamBranchTest.java]: np, made it a bit easier ;)

File: streams/src/test/java/org/apache/kafka/test/MockProcessor.java (hunk 11 above)
Comment (calmera) [reply streams/src/test/java/org/apache/kafka/test/MockProcessor.java]: I didn't look at the MockApiProcessor internals, but you are right, I can use the context from the delegate instead of tracking the context myself.

File: docs/streams/developer-guide/processor-api.html (hunk 18 above)
Comment (calmera) [reply docs/streams/developer-guide/processor-api.html]: Eventually this would be needed, but I think it is indeed best to do that as a followup tbh.

File: docs/streams/developer-guide/processor-api.html (hunk 14 above)
Comment (calmera) [reply docs/streams/developer-guide/processor-api.html]: a lot better indeed. Modified it.

File: streams/src/main/java/org/apache/kafka/streams/Topology.java (hunk 8 above)
Comment (calmera) [reply streams/src/main/java/org/apache/kafka/streams/Topology.java:763]: Updated the javadocs to reflect that we hardcode the offset reset to earliest. We might want to create another method to allow to override that, but I can't really come up with cases where you want to make it anything else than earliest.

We can add it later on if needed.

**Comment by calmera** [comment]
I agree we need to have some tests for that but that will require a bit more time and effort. I'll try to allocate some time to get that done, but if there is someone else who wants to jump in to do that, that would be great!

File: clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java [hunk 19]
```diff
@@ -54,7 +54,7 @@
  * The consumer configuration keys
  */
 public class ConsumerConfig extends AbstractConfig {
-    private static final ConfigDef CONFIG;
+    protected static final ConfigDef CONFIG;
```
Comment (mjsax) [inline clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java]: Why this change? Also what is all this `ReverseXxx` stuff below? Did you add something that does not belong to this PR by accident?

File: streams/src/main/java/org/apache/kafka/streams/Topology.java [hunk 20]
```diff
@@ -737,6 +737,85 @@ public synchronized Topology addStateStore(final StoreBuilder<?> storeBuilder,
...
+        internalTopologyBuilder.addSource(AutoOffsetReset.EARLIEST, sourceName, timestampExtractor, keyDeserializer, valueDeserializer, topic);
+        internalTopologyBuilder.addProcessor(processorName, stateUpdateSupplier, sourceName);
+        internalTopologyBuilder.addStateStore(storeBuilder, processorName);
+        internalTopologyBuilder.connectSourceStoreAndTopic(storeBuilder.name(), topic);
+
+        return this;
+    }
+
+    /**
+     * Adds a Read Only {@link StateStore} to the topology.
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/Topology.java]:
```suggestion
     * Adds a read-only {@link StateStore} to the topology.
```

File: streams/src/main/java/org/apache/kafka/streams/Topology.java [hunk 21]
```diff
@@ -737,6 +737,85 @@ public synchronized Topology addStateStore(final StoreBuilder<?> storeBuilder,
...
+     *
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/Topology.java]:
```suggestion
     * <p>
```

File: streams/src/main/java/org/apache/kafka/streams/Topology.java [hunk 22]
```diff
@@ -737,6 +737,85 @@ public synchronized Topology addStateStore(final StoreBuilder<?> storeBuilder,
         return this;
     }
 
+    /**
+     * Adds a read-only {@link StateStore} to the topology.
+     * <p>
+     * A read-only StateStore can use any compacted topic as a changelog. The <code>auto.offset.reset</code> property
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/Topology.java]:
```suggestion
     * A read-only {@link StateStore} can use any compacted topic as a changelog. The <code>auto.offset.reset</code> property
```

File: streams/src/main/java/org/apache/kafka/streams/Topology.java [hunk 23]
```diff
@@ -737,6 +737,85 @@ public synchronized Topology addStateStore(final StoreBuilder<?> storeBuilder,
...
+        internalTopologyBuilder.addStateStore(storeBuilder, processorName);
+        internalTopologyBuilder.connectSourceStoreAndTopic(storeBuilder.name(), topic);
+
+        return this;
+    }
+
+    /**
+     * Adds a Read Only {@link StateStore} to the topology.
+     *
+     * A Read Only StateStore can use any compacted topic as a changelog. The <code>auto.offset.reset</code> property
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/Topology.java]:
```suggestion
     * A read-only {@link StateStore} can use any compacted topic as a changelog. The <code>auto.offset.reset</code> property
```

**Review by mjsax — COMMENTED** [review]
I doubt that anybody will jump onto this PR quickly to help writing a test. Let me know if you will realistically get to it or not -- if not, I might try to make some time to add a test to this PR) -- I don't feel good to merge without the test, because experience shows that we might not get the test after the fact of merging....

File: clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java (hunk 19 above)
Comment (calmera) [reply clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java]: Hmm looks like my reverse consumer work got committed by accident. will clean it and go through your comments

**Comment by mjsax** [comment]
@calmera -- sorry to dropping the ball on this one -- can we resurrect it? -- First step would be rebase to resolve the conflict.

**Comment by mjsax** [comment]
@calmera -- As discussed in person (a long time back...) I took the liberty to update this PR. Hope we can get this merged soon.

File: streams/src/test/java/org/apache/kafka/test/MockProcessor.java [hunk 24]
```diff
@@ -43,12 +42,12 @@ public MockProcessor() {
     }
 
     @Override
-    public void init(ProcessorContext<KOut, VOut> context) {
+    public void init(final ProcessorContext<KOut, VOut> context) {
```
Comment (wcarlson5) [inline streams/src/test/java/org/apache/kafka/test/MockProcessor.java:45]: Thanks for cleaning up

File: streams/src/test/java/org/apache/kafka/streams/processor/ReadOnlyStoreTest.java [hunk 25]
```diff
@@ -0,0 +1,132 @@
...
+                    ));
+                }
+            },
+            "source"
+        );
+        topology.connectProcessorAndStateStores("processor", "readOnlyStore");
+        topology.addSink("sink", "outputTopic", new IntegerSerializer(), new StringSerializer(), "processor");
+
+        try (final TopologyTestDriver driver = new TopologyTestDriver(topology)) {
+            final TestInputTopic<Integer, String> changelog =
```
Comment (wcarlson5) [inline streams/src/test/java/org/apache/kafka/streams/processor/ReadOnlyStoreTest.java]: Naming this changelog and then writing to it was about 5% confusing. 

super nit but maybe 'readOnlyStoreTopicChangelog'. Idk jut something that isn't an already used term

File: streams/src/test/java/org/apache/kafka/streams/processor/ReadOnlyStoreTest.java [hunk 26]
```diff
@@ -0,0 +1,132 @@
...
+import java.util.LinkedList;
+import java.util.List;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+
+public class ReadOnlyStoreTest {
+
+    @Test
+    public void shouldLoadDataIntoReadOnlyStoreAndAllowAccessFromProcessor() {
```
Comment (wcarlson5) [inline streams/src/test/java/org/apache/kafka/streams/processor/ReadOnlyStoreTest.java]: This kind of sounds like its testing the error path if the processor shouldn't have access to it tries to write/access that. Do we have a test for that? It seems that's what's special about a read only store

**Review by wcarlson5 — APPROVED** [review]
LGTM, only a couple of nits/questions

File: streams/src/test/java/org/apache/kafka/streams/processor/ReadOnlyStoreTest.java (hunk 26 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/processor/ReadOnlyStoreTest.java]: It's not about the error path -- it's really just regular processing. Happy to rename. Suggestions? -- In the end, this test verifies that we can read data from the input topic and put into the store, plus, plus that some other Processor can access the store (we don't really have a "read-only" guard and cannot enforce it; and we can only verify if we can "connect" some other Processor to the read-only store and get a handle on it -- and we test read, as write is not allowed, even if we cannot enforce it...)

The error path, ie, store recovery from the input topic (as we don't have a changelog) is tested via new test in `RestoreIntegrationTest`.

Comment (wcarlson5) [reply streams/src/test/java/org/apache/kafka/streams/processor/ReadOnlyStoreTest.java]: Maybe shouldLoadDataIntoReadOnlyStoreFromConnectedProcessor or shouldConnectProcessorAndWriteDataToReadOnlyStore.

**Comment by mjsax** [comment]
Merged to `trunk`. Thanks for the KIP and PR @calmera!

**Comment by calmera** [comment]
Nice job @mjsax @wcarlson5! Sorry I couldn't spend more time on it.
